src/errormgr.rs

```diff
@@ -5,30 +5,69 @@ pub use self::ffi::jpeg_error_mgr as ErrorMgr;
 use self::ffi::jpeg_common_struct;
 use ::std::mem;
 
+#[repr(C)]
+pub struct WrapperError {
+    buf: *const u8,
+    len: usize,
+    cap: usize,
+}
+
+cpp!{{
+    #include<stdexcept>
+
+    struct WrapperError {
+        char* buf;
+        uintptr_t len;
+        uintptr_t cap;
+    };
+
+    struct mozjpeg_rust_wrapper_exception: std::runtime_error {
+        mozjpeg_rust_wrapper_exception(): std::runtime_error("") {}
+        WrapperError error;
+    };
+
+    extern "C" WrapperError panic_error_exit_rust(void* cinfo);
+
+    extern "C" void panic_error_exit(void* cinfo) {
+        WrapperError error = panic_error_exit_rust(cinfo);
+        mozjpeg_rust_wrapper_exception e;
+        e.error = error;
+        throw e;
+    }
+}}
```

Comment:
All this code can be automatically generated, e.g., we could have a library with a `panic_on_ffi` proc macro and a result like this:

```rust
enum MaybeFfiPanic<T> { Result(T), Panic(String) }
```

and then you would just write:

```rust
#[panic_on_ffi]
pub fn panic_error_exit_rust(cinfo: &mut jpeg_common_struct) -> MaybeFfiPanic<()> {
  MaybeFfiPanic::Panic(// ... code returning the error string ...)
}
```

and we can replace that with: 

```rust
#[no_mangle]
pub fn panic_error_exit_rust_wrapper(cinfo: &mut jpeg_common_struct) -> WrapperResult<()> {
  match {  // ... code from the function returning a string // } {
     MaybeFfiPanic::Panic(string) => WrapperResult::error(__unique_id),
     MaybeFfiPanic::Result(v) => WrapperResult::value(v),
  }
}
cpp!{{
 extern "C" WrapperResult<void> panic_error_exit_rust_wrapper(..args..);
 extern "C" void panic_error_exit(...args...) {
        WrapperResult result = panic_error_exit_rust_wrapper(...args...);
        if !result.is_error() {
            return; // return type is void  
       } else {
            mozjpeg_rust_wrapper_exception e;
            e.error = result.value.error;
            throw e;
       }
}}
#[allow(non_shouting_statics)]
static panic_error_exit: fn(...) = { 
    extern "C" { fn panic_error_exit(...args...); }
    panic_error_exit
}
```

so that you can just use the `panic_error_message` function below as usual.

We could also try to be a bit more clever, and if the return type of the Rust callback isn't  `MaybeFfiPanic`, we could auto-generate a `catch_unwind` and transform it in a `MaybeFfiPanic` for you.

Reply:
For a single callback, this feels overkill.